File: credit-engine-api/src/main/java/com/credit/engine/application/dto/currency/ExchangeRateRequest.java
package com.credit.engine.application.dto.currency;

import jakarta.validation.constraints.*;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

public record ExchangeRateRequest(
        @NotBlank(message = "O código da moeda de origem é obrigatório.")
        @Size(min = 3, max = 3, message = "O código da moeda de origem deve conter exatamente 3 letras (ex: USD).")
        String originCurrencyCode,

        @NotBlank(message = "O código da moeda de destino é obrigatório.")
        @Size(min = 3, max = 3, message = "O código da moeda de destino deve conter exatamente 3 letras (ex: BRL).")
        String destinationCurrencyCode,

        @NotNull(message = "A taxa de câmbio é obrigatória.")
        @DecimalMin(value = "0.0", inclusive = false, message = "A taxa de câmbio deve ser maior que zero.")
        BigDecimal rate,

        @NotNull(message = "A data/hora da cotação é obrigatória.")
        @PastOrPresent(message = "A data/hora da cotação não pode estar no futuro.")
        OffsetDateTime rateDateTime
) {
}

File: credit-engine-ui/src/services/currency/dto/ExchangeRateRequest.ts
import ValidationUtils from '../../../utils/ValidationUtils.ts';

export default interface ExchangeRateRequest {
    originCurrencyCode: string;
    destinationCurrencyCode: string;
    rate: number | null;
    rateDateTime: Date | null;
}

export const validateExchangeRateRequest = (request: ExchangeRateRequest): { message: string }[] => {
    const errors: { message: string }[] = [];

    if (ValidationUtils.isBlank(request.originCurrencyCode))
        errors.push({ message: 'A moeda de origem é obrigatória.' });

    if (ValidationUtils.isBlank(request.destinationCurrencyCode))
        errors.push({ message: 'A moeda de destino é obrigatória.' });

    if (request.rate === null || request.rate <= 0)
        errors.push({ message: 'A taxa de câmbio deve ser maior que zero.' });

    if (!request.rateDateTime)
        errors.push({ message: 'A data/hora da cotação é obrigatória.' });
    else if (request.rateDateTime.getTime() > Date.now())
        errors.push({ message: 'A data/hora da cotação não pode estar no futuro.' });

    return errors;
};
